program-test/src/programs.rs

```diff
@@ -21,76 +26,198 @@ mod spl_associated_token_account {
     solana_sdk::declare_id!("ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL");
 }
 
-static SPL_PROGRAMS: &[(Pubkey, Pubkey, &[u8])] = &[
-    (
-        spl_token::ID,
-        solana_sdk::bpf_loader::ID,
-        include_bytes!("programs/spl_token-3.5.0.so"),
-    ),
-    (
-        spl_token_2022::ID,
-        solana_sdk::bpf_loader_upgradeable::ID,
-        include_bytes!("programs/spl_token_2022-1.0.0.so"),
-    ),
-    (
-        spl_memo_1_0::ID,
-        solana_sdk::bpf_loader::ID,
-        include_bytes!("programs/spl_memo-1.0.0.so"),
-    ),
-    (
-        spl_memo_3_0::ID,
-        solana_sdk::bpf_loader::ID,
-        include_bytes!("programs/spl_memo-3.0.0.so"),
-    ),
-    (
-        spl_associated_token_account::ID,
-        solana_sdk::bpf_loader::ID,
-        include_bytes!("programs/spl_associated_token_account-1.1.1.so"),
-    ),
+#[derive(Error, Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ProgramLoadError {
+    #[error("Unknown program ID: {0}")]
+    UnknownId(Pubkey),
+}
+
+struct ProgramInfo {
+    program_id: Pubkey,
+    loader_id: Pubkey,
+    elf: &'static [u8],
+}
+
+static SPL_PROGRAMS: &[ProgramInfo] = &[
+    ProgramInfo {
+        program_id: spl_token::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_token-3.5.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_token_2022::ID,
+        loader_id: bpf_loader_upgradeable::ID,
+        elf: include_bytes!("programs/spl_token_2022-1.0.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_memo_1_0::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_memo-1.0.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_memo_3_0::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_memo-3.0.0.so"),
+    },
+    ProgramInfo {
+        program_id: spl_associated_token_account::ID,
+        loader_id: bpf_loader::ID,
+        elf: include_bytes!("programs/spl_associated_token_account-1.1.1.so"),
+    },
 ];
 
+/// Describes one or more accounts that belong to the same program.
+pub enum ProgramAccounts {
+    /// Non-upgradable programs, one that use the [`bpf_loader`] loader, use a single account.  This
+    /// account will just directly hold ELF of the program.
+    NonUpgradable {
+        id: Pubkey,
+        account: AccountSharedData,
+    },
+
+    /// Upgradable programs, ones that use the [`bpf_loader_upgradeable`] loader, use two accounts.
+    ///
+    /// ELF data is stored in the `program_data_account`, but it is prefixed by an additional
+    /// header.  See [`UpgradeableLoaderState::ProgramData`].
+    Upgradable {
+        main_id: Pubkey,
+        main_account: AccountSharedData,
+        program_data_id: Pubkey,
+        program_data_account: AccountSharedData,
+    },
```

Comment:
nit: I think we should stick to the naming conventions used in the upgradeable loader, ie.
- `program_address` and `program_data_address`
- `program_account` and `program_data_account`
- `Upgradeable`

Reply:
I agree we have pretty bad duality on `id` vs. `address`. I was more concerned with introducing the keyword `main`, which I don't see used anywhere else to describe program accounts, especially in the loaders.

Reply:
Ah, I just realized that I did not rename the `main` part.

Well, I think it is actually another problem that we do not have a separate name for the main account of the program.
Consider what happens if I rename `main_account` to `program_account`.
Data stored in this account is now called `program_account_data`.
But we also have a `program_data_account`.
Which is the account holding the program data.
I find it quite confusing.

If you insist, I will do it.  But I think it reduces readability.

Reply:
Yeah, I agree `program_account.data` and `program_data_account.data` aren't the greatest combination, but I think it makes more sense to align with the Loader.

https://github.com/solana-labs/solana/blob/74758d9fbfa811f7c868ad2ecca05ce5db1add25/programs/bpf_loader/src/lib.rs#L710

https://github.com/solana-labs/solana/blob/74758d9fbfa811f7c868ad2ecca05ce5db1add25/programs/bpf_loader/src/lib.rs#L769

I'd much rather use `program_data` over `programdata`, but the latter is technically the current convention, so if that's better in your opinion then it makes sense!

Reply:
Fwiw, I got hung up on `main` because it has no meaning to me in this context; I tend to agree with Joe here.